File: python/stack/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        left = {"(": ")", "{": "}", "[": "]"}
        for i in s:
            if i in left.keys():
                stack.append(i)
            else:
                if len(stack) > 0 and left[stack[-1]] == i:
                    stack.pop()
                else:
                    return False

        return len(stack) == 0

File: c++/stack/valid-parentheses.cc
class Solution {
public:
    bool isValid(string s) {
        std::unordered_map<char, char> parentheses = {{'(', ')'}, {'[', ']'}, {'{', '}'}};
        std::vector<char> left_parentheses{'(', '[', '{'};
        std::vector<char> stack;

        for (char c : s) {
            if (std::find(left_parentheses.begin(), left_parentheses.end(), c) != left_parentheses.end()) {
                stack.push_back(c);
            } else {
                if (stack.size() == 0 || parentheses[stack.back()] != c) {
                    return false;
                }
                stack.pop_back();
            }
        }

        return stack.size() == 0;
    }
};
